src/utils/roundrobin_poll_adapter.cpp

```diff
@@ -0,0 +1,161 @@
+  revocation_callback_(consumer.get_revocation_callback()),
+  rebalance_error_callback_(consumer.get_rebalance_error_callback()),
+  consumer_queue_(consumer.get_consumer_queue()) {
+    // take over the assignment callback
+    consumer_.set_assignment_callback([this](TopicPartitionList& partitions) {
+        on_assignment(partitions);
+    });
+    // take over the revocation callback
+    consumer_.set_revocation_callback([this](const TopicPartitionList& partitions) {
+        on_revocation(partitions);
+    });
+    // take over the rebalance error callback
+    consumer_.set_rebalance_error_callback([this](Error error) {
+        on_rebalance_error(error);
+    });
+}
+
+RoundRobinPollAdapter::~RoundRobinPollAdapter() {
+    restore_forwarding();
+    //set the original callbacks
+    consumer_.set_assignment_callback(assignment_callback_);
+    consumer_.set_revocation_callback(revocation_callback_);
+    consumer_.set_rebalance_error_callback(rebalance_error_callback_);
+}
+
+void RoundRobinPollAdapter::set_timeout(std::chrono::milliseconds timeout) {
+    consumer_.set_timeout(timeout);
+}
+
+std::chrono::milliseconds RoundRobinPollAdapter::get_timeout() {
+    return consumer_.get_timeout();
+}
+
+Message RoundRobinPollAdapter::poll() {
+    return poll(consumer_.get_timeout());
+}
+
+Message RoundRobinPollAdapter::poll(std::chrono::milliseconds timeout) {
+    bool empty_list = partition_queues_.ref().empty();
+    // Poll group event queue first
+    Message message = consumer_queue_.consume(empty_list ? timeout : milliseconds(0));
+    if (message) {
+        return message;
+    }
+    if (!empty_list) {
+        //consume the next partition
+        message = partition_queues_.next().consume(timeout);
+    }
+    return message;
+}
+
+MessageList RoundRobinPollAdapter::poll_batch(size_t max_batch_size) {
+    return poll_batch(max_batch_size, consumer_.get_timeout());
+}
+
+MessageList RoundRobinPollAdapter::poll_batch(size_t max_batch_size,
+                                              std::chrono::milliseconds timeout) {
+    bool empty_list = partition_queues_.ref().empty();
+    ssize_t remaining_count = max_batch_size;
+    // batch from the group event queue first
+    MessageList messages = consumer_queue_.consume_batch(remaining_count,
+                                                         empty_list ? timeout : milliseconds(0));
+    remaining_count -= messages.size();
+    if ((remaining_count <= 0) || empty_list) {
+        // the entire batch was filled
+        return messages;
+    }
+    // batch from the next partition
+    MessageList partition_messages = partition_queues_.next().consume_batch(remaining_count, timeout);
+    if (messages.empty()) {
+        return partition_messages;
+    }
+    if (partition_messages.empty()) {
+        return messages;
+    }
+    // concatenate both lists
+    messages.reserve(messages.size() + partition_messages.size());
```

Comment:
I use reserve for performance reasons and I think we should keep it as the batch size could be quite large. That's the reason I added the return line above the reserve, in case it's empty there's not need to go forward.

Reply:
Insert already expands the vector (once) so this call is useless. `insert` knows how many elements there are between the two you provide so it knows how much to enlarge the vector.